base_layer/core/src/base_node/sync/header_sync/synchronizer.rs

```diff
@@ -380,69 +398,76 @@ impl<'a, B: BlockchainBackend + 'static> HeaderSynchronizer<'a, B> {
-            .await?;
-        let proto::FindChainSplitResponse {
-            headers,
-            fork_hash_index,
-            tip_height: remote_tip_height,
-        } = resp;
-
-        if steps_back > 0 {
+        sync_peer: &str,
+        best_header: ChainHeader,
+        best_block: ChainHeader,
+        peer_response: PeerChainSplitResponse,
+    ) -> Result<HeaderSyncStatus, BlockHeaderSyncError> {
+        if peer_response.reorg_steps_back > 0 {
             debug!(
                 target: LOG_TARGET,
                 "Found chain split {} blocks back, received {} headers from peer `{}`",
-                steps_back,
-                headers.len(),
+                peer_response.reorg_steps_back,
+                peer_response.headers.len(),
                 sync_peer
             );
         }
 
-        // If the peer returned no new headers, this means header sync is done.
-        if headers.is_empty() {
-            if fork_hash_index > 0 {
+        // If the peer returned no new headers, they have no headers, but may still have more blocks than we have,
+        // thus have a higher accumulated difficulty.
+        if peer_response.headers.is_empty() {
+            if peer_response.fork_hash_index > 0 {
                 debug!(
                     target: LOG_TARGET,
-                    "Peer `{}` has sent no headers but forked_hash_index is {}. The peer is behind our chain.",
+                    "Peer `{}` has sent no headers with forked_hash_index {}. The peer has less headers than we have.",
                     sync_peer,
-                    fork_hash_index
+                    peer_response.fork_hash_index
                 );
-
-                return Ok(SyncStatus::WereAhead);
+            } else {
+                debug!(target: LOG_TARGET, "Headers already in sync with peer `{}`.", sync_peer);
             }
 
-            debug!(target: LOG_TARGET, "Already in sync with peer `{}`.", sync_peer);
-            return Ok(SyncStatus::InSync);
+            return Ok(HeaderSyncStatus::InSyncOrAhead);
         }
 
-        let headers = headers
+        let headers = peer_response
+            .headers
             .into_iter()
             .map(BlockHeader::try_from)
             .collect::<Result<Vec<_>, _>>()
             .map_err(BlockHeaderSyncError::ReceivedInvalidHeader)?;
-        let num_new_headers = headers.len();
+        let num_new_headers = headers.len(); // Required for later use, no 'Copy' trait on 'BlockHeader'
 
         // NOTE: We can trust that the header associated with this hash exists because `block_hashes` was supplied by
         // this node. Bounds checking for fork_hash_index has been done above.
         #[allow(clippy::cast_possible_truncation)]
-        let chain_split_hash = block_hashes.get(fork_hash_index as usize).unwrap();
+        let chain_split_hash = peer_response
+            .block_hashes
+            .get(peer_response.fork_hash_index as usize)
```

Comment:
This unwrap seems like the peer could cause a panic if `fork_hash_index > block_hashes.len()`

Reply:
That is not possible, it is checked above, with a comment git is not showing
https://github.com/tari-project/tari/blob/6e52f447a5f127af4a7523fb0f26684bc9e9173b/base_layer/core/src/base_node/sync/header_sync/synchronizer.rs#L442C11-L443C79

Reply:
Simplified - `find_chain_split` will return the `chain_split_hash `